File: packages/sdkwork-utils-php/src/CollectionUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class CollectionUtils
{
    public static function unique(array $items): array
    {
        $seen = [];
        $result = [];
        foreach ($items as $item) {
            $key = is_object($item) ? spl_object_hash($item) : (string) json_encode($item);
            if (isset($seen[$key])) {
                continue;
            }
            $seen[$key] = true;
            $result[] = $item;
        }
        return $result;
    }

    public static function chunk(array $items, int $size): array
    {
        if ($size <= 0) {
            return [];
        }
        return array_chunk($items, $size);
    }

    public static function groupBy(array $items, callable $keyFn): array
    {
        $groups = [];
        foreach ($items as $item) {
            $groups[$keyFn($item)][] = $item;
        }
        return $groups;
    }

    public static function flatten(array $items): array
    {
        $result = [];
        foreach ($items as $group) {
            foreach ($group as $item) {
                $result[] = $item;
            }
        }
        return $result;
    }

    public static function compact(array $items): array
    {
        return array_values(array_filter($items, static fn ($item) => $item !== null));
    }

    public static function first(array $items): mixed
    {
        return $items === [] ? null : $items[0];
    }

    public static function last(array $items): mixed
    {
        return $items === [] ? null : $items[array_key_last($items)];
    }

    public static function sortBy(array $items, callable $keyFn): array
    {
        usort($items, static function ($left, $right) use ($keyFn) {
            return $keyFn($left) <=> $keyFn($right);
        });
        return $items;
    }

    public static function keyBy(array $items, callable $keyFn): array
    {
        $result = [];
        foreach ($items as $item) {
            $result[$keyFn($item)] = $item;
        }
        return $result;
    }

    public static function filter(array $items, callable $predicate): array
    {
        return array_values(array_filter($items, $predicate));
    }

    public static function find(array $items, callable $predicate): mixed
    {
        foreach ($items as $item) {
            if ($predicate($item)) {
                return $item;
            }
        }
        return null;
    }
}

File: packages/sdkwork-utils-csharp/CollectionUtils.cs
namespace Sdkwork.Utils;

public static class CollectionUtils
{
    public static List<T> Unique<T>(IEnumerable<T> items)
    {
        var seen = new HashSet<T>();
        var result = new List<T>();
        foreach (var item in items)
        {
            if (seen.Add(item)) result.Add(item);
        }
        return result;
    }

    public static List<List<T>> Chunk<T>(IReadOnlyList<T> items, int size)
    {
        if (size <= 0) return [];
        var result = new List<List<T>>();
        for (var index = 0; index < items.Count; index += size)
        {
            result.Add(items.Skip(index).Take(size).ToList());
        }
        return result;
    }

    public static Dictionary<TKey, List<T>> GroupBy<T, TKey>(
        IEnumerable<T> items,
        Func<T, TKey> keyFn) where TKey : notnull
    {
        var groups = new Dictionary<TKey, List<T>>();
        foreach (var item in items)
        {
            var key = keyFn(item);
            if (!groups.TryGetValue(key, out var group))
            {
                group = [];
                groups[key] = group;
            }
            group.Add(item);
        }
        return groups;
    }

    public static List<T> Flatten<T>(IEnumerable<IEnumerable<T>> items) => items.SelectMany(item => item).ToList();

    public static List<T> Compact<T>(IEnumerable<T?> items) where T : class =>
        items.Where(item => item is not null).Cast<T>().ToList();

    public static object? First<T>(IReadOnlyList<T> items) =>
        items.Count > 0 ? items[0] : null;

    public static object? Last<T>(IReadOnlyList<T> items) =>
        items.Count > 0 ? items[^1] : null;

    public static List<T> SortBy<T, TKey>(IEnumerable<T> items, Func<T, TKey> keyFn)
        where TKey : IComparable<TKey> =>
        items.OrderBy(keyFn).ToList();

    public static Dictionary<TKey, T> KeyBy<T, TKey>(IEnumerable<T> items, Func<T, TKey> keyFn)
        where TKey : notnull
    {
        var result = new Dictionary<TKey, T>();
        foreach (var item in items)
        {
            result[keyFn(item)] = item;
        }

        return result;
    }

    public static List<T> Filter<T>(IEnumerable<T> items, Func<T, bool> predicate) =>
        items.Where(predicate).ToList();

    public static T? Find<T>(IEnumerable<T> items, Func<T, bool> predicate) =>
        items.FirstOrDefault(predicate);
}
